src/statement.ts

```diff
@@ -8,122 +9,344 @@ import { BaseNode, isNode, typeGuard } from "./node";
 export type Stmt =
   | BreakStmt
   | BlockStmt
+  | CatchClause
+  | ContinueStmt
+  | DoStmt
   | ExprStmt
   | ForInStmt
   | ForOfStmt
   | IfStmt
   | ReturnStmt
-  | VariableStmt;
+  | ThrowStmt
+  | TryStmt
+  | VariableStmt
+  | WhileStmt;
 
 export function isStmt(a: any): a is Stmt {
   return (
     isNode(a) &&
-    (isBlock(a) ||
+    (isBreakStmt(a) ||
+      isBlockStmt(a) ||
+      isCatchClause(a) ||
+      isContinueStmt(a) ||
       isExprStmt(a) ||
       isForInStmt(a) ||
       isForOfStmt(a) ||
       isIfStmt(a) ||
       isReturn(a) ||
+      isThrowStmt(a) ||
+      isTryStmt(a) ||
       isVariableStmt(a))
   );
 }
 
+export abstract class BaseStmt<
+  Kind extends FunctionlessNode["kind"],
+  Parent extends FunctionlessNode | undefined = BlockStmt | IfStmt
+> extends BaseNode<Kind, Parent> {
+  readonly nodeKind: "Stmt" = "Stmt";
+
+  /**
+   * Node that is prior to this node.
+   */
+  prev: Stmt | undefined;
+  /**
+   * Node that is subsequent to this node.
+   */
+  next: Stmt | undefined;
+}
+
 export const isExprStmt = typeGuard("ExprStmt");
 
-export class ExprStmt extends BaseNode<"ExprStmt"> {
+export class ExprStmt extends BaseStmt<"ExprStmt"> {
   constructor(readonly expr: Expr) {
     super("ExprStmt");
-    expr.parent = this;
+    expr.setParent(this);
+  }
+
+  public clone(): this {
+    return new ExprStmt(this.expr.clone()) as this;
   }
 }
 
 export const isVariableStmt = typeGuard("VariableStmt");
 
+export type VariableStmtParent =
+  | ForInStmt
+  | ForOfStmt
+  | FunctionDecl
+  | FunctionExpr
+  | CatchClause;
+
 export class VariableStmt<
   E extends Expr | undefined = Expr | undefined
-> extends BaseNode<"VariableStmt"> {
+> extends BaseStmt<"VariableStmt", VariableStmtParent> {
   constructor(readonly name: string, readonly expr: E) {
     super("VariableStmt");
     if (expr) {
-      expr.parent = this;
+      expr.setParent(this);
     }
   }
+
+  public clone(): this {
+    return new VariableStmt(this.name, this.expr?.clone()) as this;
+  }
 }
 
-export const isBlock = typeGuard("BlockStmt");
+export const isBlockStmt = typeGuard("BlockStmt");
 
-export class BlockStmt extends BaseNode<"BlockStmt"> {
+export type BlockStmtParent =
+  | CatchClause
+  | DoStmt
+  | ForInStmt
+  | ForOfStmt
+  | FunctionDecl
+  | FunctionExpr
+  | IfStmt
+  | TryStmt
+  | WhileStmt;
+
+export class BlockStmt extends BaseStmt<"BlockStmt", BlockStmtParent> {
   constructor(readonly statements: Stmt[]) {
     super("BlockStmt");
-    let prev = undefined;
-    for (const expr of statements) {
-      expr.parent = this;
-      expr.prev = prev;
-      prev = expr;
+    statements.forEach((stmt, i) => {
+      stmt.setParent(this as never);
+      stmt.prev = i > 0 ? statements[i - 1] : undefined;
+      stmt.next = i + 1 < statements.length ? statements[i + 1] : undefined;
+    });
```

Comment:
make `prev` and `next` a parameter of stmt? I don't like the need to pass into a Block in order for these magic properties to be set.

Reply:
This is the cleanest way to enable the tree to be created naturally:
```ts
new FunctionDecl([], new BlockStmt([
  new ExprStmt(..),
  ..
]))
```

Reply:
the definition of `prev` and `next` is owned by `BlockStmt` - no other node should ever have to worry about it. The only condition is that a Node is used only once. Use `.clone()` if you need to put it somewhere else.